pkg/operator/controller/dns/controller.go

```diff
@@ -222,13 +223,26 @@ func (r *reconciler) createDNSProviderIfNeeded(dnsConfig *configv1.DNS) error {
 	creds := &corev1.Secret{}
 	switch platformStatus.Type {
 	case configv1.AWSPlatformType, configv1.AzurePlatformType, configv1.GCPPlatformType:
-		name := types.NamespacedName{Namespace: r.config.Namespace, Name: cloudCredentialsSecretName}
-		if err := r.cache.Get(context.TODO(), name, creds); err != nil {
-			return fmt.Errorf("failed to get cloud credentials from secret %s: %v", name, err)
+		if (r.infraConfig != nil && infraConfig.Status.ControlPlaneTopology != configv1.ExternalTopologyMode) || r.infraConfig == nil {
+			name := types.NamespacedName{Namespace: r.config.Namespace, Name: cloudCredentialsSecretName}
+			if err := r.cache.Get(context.TODO(), name, creds); err != nil {
+				return fmt.Errorf("failed to get cloud credentials from secret %s: %v", name, err)
+			}
+
+			if r.cloudCredentials == nil || !reflect.DeepEqual(creds.Data, r.cloudCredentials.Data) {
+				needUpdate = true
+			}
 		}
+	case configv1.IBMCloudPlatformType:
+		if r.infraConfig != nil && infraConfig.Status.ControlPlaneTopology != configv1.ExternalTopologyMode {
```

Comment:
In https://github.com/openshift/cluster-ingress-operator/compare/634e4b2c2a6f74899a3e310bc0c5c28d449d611d..7000e8a956b886d99353c456ddbb4a2744389b2f, you deleted this separate `case` branch and added `configv1.IBMCloudPlatformType` to the `case` above, and then in https://github.com/openshift/cluster-ingress-operator/compare/75f62202398fee26e230891ec54bc2878f67ba88..ca476de7671bcaadaef41712615bc9db0361d29f, you added this `case` back with this condition.  I don't see why we need the two separate `case` branches.  Is it necessary to skip initialization of the DNS provider on IBM Cloud when `r.infraConfig == nil`?  We should avoid special cases where possible (and it would be nice to avoid the code duplication as well if the special casing isn't required).